### PR TITLE
Update dll.hpp with WinAPI declarations required by Boost.DLL library

### PR DESCRIPTION
**Pull request by apolukhin**
The file is copied from [Boost.DLL/include/boost/detail/winapi/dll2.hpp](https://github.com/apolukhin/Boost.DLL/blob/master/include/boost/detail/winapi/dll2.hpp)

### CODE REVIEW

File: include/boost/detail/winapi/dll.hpp [hunk 1]
```diff
@@ -1,7 +1,8 @@
 //  dll.hpp  --------------------------------------------------------------//
 
 //  Copyright 2010 Vicente J. Botet Escriba
-//  Copyright 2015 Andrey Semashev
```
Comment (Lastique) [inline include/boost/detail/winapi/dll.hpp]: Please, don't remove or replace copyright lines but only add new ones. Even if the year is the same.

File: include/boost/detail/winapi/dll.hpp [hunk 2]
```diff
@@ -18,83 +19,186 @@
...
 using ::LoadLibraryW;
+using ::LoadLibraryExW;
 using ::GetModuleHandleW;
+using ::GetModuleFileNameW;
 using ::FreeLibrary;
 using ::GetProcAddress;
+using ::VirtualQuery;
 
 #if !defined( BOOST_NO_ANSI_APIS )
-BOOST_FORCEINLINE HMODULE_ load_library(LPCSTR_ lpFileName)
```
Comment (Lastique) [inline include/boost/detail/winapi/dll.hpp]: Please, don't change the coding style of the existing code. I'd rather prefer the new code to follow the existing style.

File: include/boost/detail/winapi/dll.hpp [hunk 3]
```diff
@@ -18,83 +19,186 @@
...
+
+const DWORD_ DONT_RESOLVE_DLL_REFERENCES_           = DONT_RESOLVE_DLL_REFERENCES;
+const DWORD_ LOAD_IGNORE_CODE_AUTHZ_LEVEL_          = LOAD_IGNORE_CODE_AUTHZ_LEVEL;
+const DWORD_ LOAD_LIBRARY_AS_DATAFILE_              = LOAD_LIBRARY_AS_DATAFILE;
+const DWORD_ LOAD_LIBRARY_AS_DATAFILE_EXCLUSIVE_    = LOAD_LIBRARY_AS_DATAFILE_EXCLUSIVE;
+const DWORD_ LOAD_LIBRARY_AS_IMAGE_RESOURCE_        = LOAD_LIBRARY_AS_IMAGE_RESOURCE;
+const DWORD_ LOAD_WITH_ALTERED_SEARCH_PATH_         = LOAD_WITH_ALTERED_SEARCH_PATH;
 #endif // defined( BOOST_USE_WINDOWS_H )
 
+// We always define the following constants by explicit values, because
```
Comment (Lastique) [inline include/boost/detail/winapi/dll.hpp]: If these constants are only supported from Windows 8, their definition should probably be conditioned on `BOOST_USE_WINAPI_VERSION`, shouldn't it?

File: include/boost/detail/winapi/dll.hpp [hunk 4]
```diff
@@ -18,83 +19,186 @@
...
 BOOST_SYMBOL_IMPORT boost::detail::winapi::FARPROC_ WINAPI
 GetProcAddress(boost::detail::winapi::HMODULE_ hModule, boost::detail::winapi::LPCSTR_ lpProcName);
-}
-#endif
 
-namespace boost {
-namespace detail {
-namespace winapi {
+BOOST_SYMBOL_IMPORT boost::detail::winapi::FARPROC_ WINAPI
+VirtualQuery(
```
Comment (Lastique) [inline include/boost/detail/winapi/dll.hpp]: This function is most probably declared incorrectly (I haven't compared it to the actual declaration in Windows SDK, but fairly sure still). The function must use the same types in the signature as it does in Windows SDK, and it surely doesn't reference `boost::detail::winapi::MEMORY_BASIC_INFORMATION32_` or `boost::detail::winapi::MEMORY_BASIC_INFORMATION64_` there. This header should forward-declare the structure/union/whatever `MEMORY_BASIC_INFORMATION` exactly as it is presented in Windows SDK and use a pointer to it in the `VirtualQuery` forward declaration. `boost::detail::winapi::VirtualQuery` should be an inline function that does a `reinterpret_cast` of the `boost::detail::winapi::MEMORY_BASIC_INFORMATION_` pointer to the pointer to the real `MEMORY_BASIC_INFORMATION`.

File: include/boost/detail/winapi/dll.hpp [hunk 5]
```diff
@@ -18,83 +19,186 @@
 
 #if !defined( BOOST_USE_WINDOWS_H )
 extern "C" {
+
 namespace boost { namespace detail { namespace winapi {
+struct MEMORY_BASIC_INFORMATION64_ {
```
Comment (Lastique) [inline include/boost/detail/winapi/dll.hpp]: Does Windows SDK define both these structures? Shouldn't there be just one structure defined differently for different bitness?

**Comment by apolukhin** [comment]
> If these constants are only supported from Windows 8, their definition should probably be conditioned on BOOST_USE_WINAPI_VERSION, shouldn't it?

This is a tricky question. They are ready out-of-the-box in Windows8, but previous versions of Windows with KB2533623 update also could use it. There's no right way to check those flags at compile time and MSDN recommends some runtime tricks to ensure that those falgs do work.

I've just made sure that they are not used in Boost.DLL any more and removed them...

> Does Windows SDK define both these structures? Shouldn't there be just one structure defined differently for different bitness?

Actually Windows SDK defines 3 different structures: `_MEMORY_BASIC_INFORMATION`, `_MEMORY_BASIC_INFORMATION32`, `_MEMORY_BASIC_INFORMATION64`. `VirtualQuery` always uses `_MEMORY_BASIC_INFORMATION` whose representation depends on bitness.

Left the `boost::detail::winapi::MEMORY_BASIC_INFORMATION_` structure only.

All the other issues were fixed according to your requirements.

**Comment by Lastique** [comment]
Committed a slightly corrected version of this PR as https://github.com/boostorg/winapi/commit/9ecf1c1b1705e16b64689e85ccca59b5b1dda610.

**Comment by apolukhin** [comment]
Thanks, works well.

**Pull request closed without merging**